docs/source/schema-directives.md

```diff
@@ -1,179 +1,581 @@
+
+By overriding methods like `visitObject`, a subclass of `SchemaDirectiveVisitor` expresses interest in certain schema types such as `GraphQLObjectType` (the first parameter type of `visitObject`).
+
+When `SchemaDirectiveVisitor.visitSchemaDirectives` is called with a `GraphQLSchema` object and a map of visitor subclasses (`{ [directiveName: string]: typeof SchemaDirectiveVisitor }`), visitor methods overridden by those subclasses will be invoked with references to any schema objects that have appropriately named directives attached to them, enabling the visitors to inspect or modify the schema.
+
+For example, here is one possible implementation of the `@deprecated` directive we saw above:
+
+```typescript
+import { SchemaDirectiveVisitor } from "graphql-tools";
+
+class DeprecatedDirective extends SchemaDirectiveVisitor {
+  public visitFieldDefinition(field: GraphQLField<any, any>) {
+    value.isDeprecated = true;
+    value.deprecationReason = this.args.reason;
+  }
+
+  public visitEnumValue(value: GraphQLEnumValue) {
+    value.isDeprecated = true;
+    value.deprecationReason = this.args.reason;
+  }
+}
+```
+
+In order to apply this implementation to a schema that contains `@deprecated` directives, simply pass the `DeprecatedDirective` class to the `makeExecutableSchema` function via the `directives` option:
+
+```typescript
+import { makeExecutableSchema } from "graphql-tools";
+
+const typeDefs = `
+type ExampleType {
+  newField: String
+  oldField: String @deprecated(reason: "Use \`newField\`.")
+}`;
+
+const schema = makeExecutableSchema({
+  typeDefs,
+  directives: {
+    deprecated: DeprecatedDirective
+  }
+});
+```
+
+Alternatively, if you want to modify an existing schema object, you can use the `SchemaDirectiveVisitor.visitSchemaDirectives` interface directly:
+
+```typescript
+SchemaDirectiveVisitor.visitSchemaDirectives(schema, {
+  deprecated: DeprecatedDirective
+});
+```
+
+Note that a subclass of `SchemaDirectiveVisitor` may be instantiated multiple times to visit multiple different occurrences of the `@deprecated` directive. That's why you provide a class rather than an instance of that class.
+
+If for some reason you have a schema that uses another name for the `@deprecated` directive, that's not a problem. The same `DeprecatedDirective` class can be used with a different name, simply by changing its key in the `directives` object passed to `makeExecutableSchema`. In other words, `SchemaDirectiveVisitor` implementations are effectively anonymous, so it's up to whoever uses them to assign names to them.
+
+## Examples
+
+To appreciate the range of possibilities enabled by `SchemaDirectiveVisitor`, let's examine a variety of practical examples.
+
+> Note that these examples are written in JavaScript rather than TypeScript, though either language should work.
+
+### Uppercasing strings
+
+Suppose you want to ensure a string-valued field is converted to uppercase:
```

Comment:
I'd call out that this is an example of wrapping a resolver.

Also, fun fact: This directive won't work on fields that return an array of values, but maybe that's reasonable since it would only be expected to work on `String` fields.